nengo/builder/builder.py

```diff
@@ -82,11 +88,13 @@ def __init__(self, dt=0.001, label=None, decoder_cache=None, builder=None):
         self.seeded = {}
 
         self.sig = collections.defaultdict(dict)
-        self.sig['common'][0] = Signal(0., readonly=True, name='ZERO')
-        self.sig['common'][1] = Signal(1., readonly=True, name='ONE')
+        self.sig['common'][0] = Signal(
+            np.array(0., dtype=rc.sim_dtype), readonly=True, name='ZERO')
```

Comment:
This is what was in the commit before my squashes, to a certain extent. The reason I removed it is to be slightly more efficient -- what's the point of making a float64 array and then immediately making a new float32 copy? It would be different if it was something the user was passing us, but we are responsible for making these numpy array, so we might as well make setting the dtype a one-step process instead of a two-step process.

Of course, the way it was done before still required modifying all of these signals, so there was no benefit in terms of reducing the diff. Doing it like you suggest, of having `Signal.__init__` have a default, would cut down the diff, so I can see more of an argument for it, but it would still be slightly inefficient, so since I've gone through the effort of changing all of our dtypes, I'd prefer keeping it this way.

Unfortunately I can't think of a way to automatically detect when we're going to use the wrong dtype. There are legitimate reasons to use int dtypes (notable indices, which we do in a couples places) so we can't always say "hey, this should be a float". Then by the same logic, perhaps there are reasons why you might want a float64 for some operation regardless of the global dtype.

One alternative that comes to mind that would be easier to enforce would be to ask the user for a precision in bits rather than to ask for a dtype. That would make the implementation a little bit easier, and we wouldn't have to worry about super weird dtypes (which are already difficult to specify since we ask for a string that we pass to `np.dtype`). I'd be more comfortable doing the cast in `Signal.__init__` if we did it that way (though I would still keep all of these lines so that we don't make two arrays unnecessarily; casting in `Signal.__init__` would just be in case we missed it somewhere or if a user makes their own build function and forgets it there). But in that case we might also need to add something to `Signal` to differentiate float vs int signals?

Reply:
Moreso than reducing the diff, what strikes me is removing the repetitive `dtype=rc.sim_dtype` all throughout the code.  Whenever I see something like that it makes me think there has to be a way to avoid that 😄.  

The other, more significant, advantage is easing future maintainability.  In the current system, whenever we add a new Signal we need to remember to set `dtype=rc.sim_dtype`.  I am fairly certain we will not remember to do that every time, and then we'll end up with simulations with mixed dtypes.  And typically when we're adding a new Signal it's because we're adding a new feature/op, so it wouldn't be caught by the test that builds a model out of our existing ops and checks all their dtypes.  So again, maybe we will all remember to update that test whenever we add a new op, but I suspect not 🤷‍♂ .

> Unfortunately I can't think of a way to automatically detect when we're going to use the wrong dtype. There are legitimate reasons to use int dtypes (notable indices, which we do in a couples places) so we can't always say "hey, this should be a float". Then by the same logic, perhaps there are reasons why you might want a float64 for some operation regardless of the global dtype.

I wasn't thinking of automatic detection, I was thinking of someone accidentally passing an `initial_value` that doesn't match the specified `dtype` for the Signal.  So, e.g., you specify int and accidentally pass a float, or explicitly want `float64` but accidentally use `rc.sim_dtype`.  It's a minor benefit, since we usually create the initial value at the same time as the Signal so those kinds of errors are unlikely, but it wouldn't hurt.

Overall I suspect the cost of casting within the Signal init would be pretty minor (in terms of time and memory), since it's just a one-time cost during build time.  But if we did want to avoid that, we could just make the `initial_value` argument optional and add a `shape` parameter, so the Signal data would be specified abstractly with dtype/shape, and then the we'd call `np.zeros` (which we use in most cases) inside init.

Reply:
The reason we pass a matrix rather than a shape is for signals whose initial values are not zero, e.g. weights, biases, etc. We could do it so that we pass a shape and an optional "init" value, but then I'd be worried for things like weights, we don't want to make an unnecessary extra copy if we can avoid it.

Personally, I think the cost with re-casting an array is very minimal, and it's not worth putting a lot of our dev time into avoiding it. Particularly, because I think with the large signals in this commit (i.e. weights), they should already be the right dtype due to having the right dtype on things passed in to the solver.

On that note, with these changes, we should be explicitly casting transforms to the right dtype in `builder/transforms.py`. I think in many cases the value returned by `transform.sample` will be right, but we need to make sure.

Reply:
Yeah I wouldn't want to remove the ability to manually specify an `initial_value` for those reasons, this would just be an optional feature to avoid the re-casting in the common cases.  But agreed that I'm not sure how much of a difference it would actually make.